pkg/miscellaneous/miscellaneous.go

```diff
@@ -541,12 +542,18 @@ func ReadConfigFileAsMap(path string) (map[string]interface{}, error) {
 
 // ParsePeerURL forms a PeerURL, given podName by parsing the initial-advertise-peer-urls
 func ParsePeerURL(initialAdvertisePeerURLs, podName string) (string, error) {
-	tokens := strings.Split(initialAdvertisePeerURLs, "@")
-	if len(tokens) < 4 {
-		return "", fmt.Errorf("invalid peer URL : %s", initialAdvertisePeerURLs)
-	}
-	domaiName := fmt.Sprintf("%s.%s.%s", tokens[1], tokens[2], "svc")
-	return fmt.Sprintf("%s://%s.%s:%s", tokens[0], podName, domaiName, tokens[3]), nil
+	var scheme, peerSvcName, namespace, peerPort string
+	re := regexp.MustCompile(`^([^:]+)://([^\.]+)\.([^:]+):(.+)$`)
+	tokens := re.FindStringSubmatch(initialAdvertisePeerURLs)
+	if len(tokens) != 5 {
+		return "", fmt.Errorf("invalid URL format: %s", initialAdvertisePeerURLs)
+	}
+	scheme = tokens[1]
+	peerSvcName = tokens[2]
+	namespace = tokens[3]
+	peerPort = tokens[4]
+	domainName := fmt.Sprintf("%s.%s.%s", peerSvcName, namespace, "svc")
+	return fmt.Sprintf("%s://%s.%s:%s", scheme, podName, domainName, peerPort), nil
```

Comment:
The intent of the functionality is clearer to the reader when we perform the parsing manually instead of using regex.

If the input is malformed, the user would see more directed error messages like `the hostname for the initialAdvertisePeerURLs must be of the form "peerServiceName.namespace"`, and `the peer port must be specified` instead of getting a regex error.

What are your opinions on this way of doing it?

```suggestion
	u, err := url.Parse(initialAdvertisePeerURLs)
	if err != nil {
		return "", err
	}

	hostParts := strings.Split(u.Hostname(), ".")
	if len(hostParts) != 2 {
		return "", fmt.Errorf("the hostname for the initialAdvertisePeerURLs must be of the form \"peerServiceName.namespace\"")
	}
	peerSvcName, ns := hostParts[0], hostParts[1]

	if u.Port() == "" {
		return "", fmt.Errorf("the peer port must be specified")
	}

	domainName := fmt.Sprintf("%s.%s.%s", peerSvcName, ns, "svc")
	return fmt.Sprintf("%s://%s.%s:%s", u.Scheme, podName, domainName, u.Port()), nil
```

Reply:
Addressed [96818c](https://github.com/gardener/etcd-backup-restore/pull/715/commits/96818c894117b3f67791590419f5f4cd6a569583)